crates/sui-bridge/src/config.rs

```diff
@@ -0,0 +1,267 @@
+    /// Path of the client storage. Required when `run_client` is true.
+    pub db_path: Option<PathBuf>,
+    /// The eth addresses (hex) for client to watch for. Need to contain at least one item when `run_client` is true.
+    pub eth_addresses: Option<Vec<String>>,
+    /// The sui modules of bridge packages for client to watch for. Need to contain at least one item when `run_client` is true.
+    pub sui_bridge_modules: Option<Vec<String>>,
+    /// Override the start block number for each eth address. Key must be in `eth_addresses`.
+    /// When set, EthSyncer will start from this block number instead of the one in storage.
+    pub eth_addresses_start_block_number_override: Option<BTreeMap<String, u64>>,
+    /// Override the start transaction digest for each bridge module. Key must be in `sui_bridge_modules`.
+    /// When set, SuiSyncer will start from this transaction digest instead of the one in storage.
+    pub sui_bridge_modules_start_tx_override: Option<BTreeMap<String, String>>,
+}
+
+impl Config for BridgeNodeConfig {}
+
+impl BridgeNodeConfig {
+    pub async fn validate(
+        &self,
+    ) -> anyhow::Result<(BridgeServerConfig, Option<BridgeClientConfig>)> {
+        let bridge_authority_key =
+            read_bridge_authority_key(&self.bridge_authority_key_path_base64_raw)?;
+
+        // TODO: verify it's part of bridge committee
+
+        let sui_client = Arc::new(SuiClient::<SuiSdkClient>::new(&self.sui_rpc_url).await?);
+        let eth_client =
+            Arc::new(EthClient::<ethers::providers::Http>::new(&self.eth_rpc_url).await?);
+
+        let bridge_server_config = BridgeServerConfig {
+            key: bridge_authority_key,
+            metrics_port: self.metrics_port,
+            server_listen_port: self.server_listen_port,
+            sui_client: sui_client.clone(),
+            eth_client: eth_client.clone(),
+        };
+
+        if !self.run_client {
+            return Ok((bridge_server_config, None));
+        }
+        // If client is enabled, prepare client config
+        let bridge_client_key = if self.bridge_client_key_path_base64_sui_key.is_none() {
+            let bridge_client_key =
+                read_bridge_authority_key(&self.bridge_authority_key_path_base64_raw)?;
+            Ok(SuiKeyPair::from(bridge_client_key))
+        } else {
+            read_bridge_client_key(self.bridge_client_key_path_base64_sui_key.as_ref().unwrap())
+        }?;
+
+        let client_sui_address = SuiAddress::from(&bridge_client_key.public());
+        let gas_object_id = self.bridge_client_gas_object.ok_or(anyhow!(
+            "`bridge_client_gas_object` is required when `run_client` is true"
+        ))?;
+        let db_path = self
+            .db_path
+            .clone()
+            .ok_or(anyhow!("`db_path` is required when `run_client` is true"))?;
+        if self.eth_addresses.is_none() || self.eth_addresses.as_ref().unwrap().is_empty() {
```

Comment:
think you check this again below in the match stmt, so this can be removed?

Reply:
good catch!